ERCS/erc-7771.md

```diff
@@ -0,0 +1,647 @@
+				}
+
+				if (implementation == address(0)) {
+				    revert UnknownSelector(sig4);
+				}
+
+        // Delegatecall to the implementation contract
+        assembly {
+            calldatacopy(0, 0, calldatasize())
+
+            let result := delegatecall(gas(), implementation, 0, calldatasize(), 0, 0)
+            returndatacopy(0, 0, returndatasize())
+
+            switch result
+            case 0 {
+                revert(0, returndatasize())
+            }
+            default {
+                return(0, returndatasize())
+            }
+        }
+    }
+}
+```
+
+### Router Code Generation
+
+Router code SHOULD NOT be written manually, and generated by automated tooling instead. Tooling MUST ensure unique function selectors across all modules and that there are no storage collisions between any of its modules.
+
+The binary search algorithm used in the lookup table SHOULD order function selectors numerically, and cluster selectors by an optimal number of 5-10 selectors. This ensures that the routing gas cost of calling any selector in the system is consistent.
+
+### Upgradeability
+
+A router does not inherently make a system upgradeable.
+
+However, a router can be used in an upgradeable system by using it as the implementation of a single system proxy. Such a proxy does not need admin upgradeability functions, since upgradeability can be delegated to an `UpgradeModule` that defines upgrade functionality.
+
+### Function Hinting
+
+Interfaces can be automatically generated to interact with protocols employing the router proxy by scanning the router’s code for `@module` tokens, which provide a module name and an implementation address.
+
+## Rationale
+
+### Overview
+
+In a regular proxy, an incoming call to function `b()` is forwarded using `DELEGATECALL` to the proxy’s current implementation.
+![2](../assets/eip-7771/images/diag2.jpeg)
+
+The router proxy is similar to a regular proxy, in that it receives incoming calls and forwards them to an implementation, the only difference being that a router may forward the call one of many —simultaneous— implementation contracts. It decides which implementation to forward to using a hardcoded lookup table that maps function selectors to their corresponding implementation contracts, without the need of using storage.
+
+![3](../assets/eip-7771/images/diag3.jpeg)
+
+In the example above, the lookup table is used to determine that the selector for the function `b()` is implemented in the address of `ModuleB`.
+
+This enables the router to overcome the 24KB contract size limit, since this limitation is now only relevant within each module, thus enabling the router to practically “merge” all the modules’ code.
+
+The only caveat is that function selectors must be unique in all modules. I.e. `ModuleC` may not define a function `b()` if `ModuleB` also defines `b()`.
+
+Let’s continue, by looking at how the hardcoded lookup table is implemented.
+
+### Router Implementation
```

Comment:
Very little of this section is describing the rationale behind decisions made within the proposal. Trim this section down. Instructions to implementors don't really belong in the standard itself.